### PR TITLE
Adding device and dtype to a number of module factories.

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
Renamed a few parameters to align with PyTorch.
Added parameters and buffers to InstanceNorm{123}d.

### CODE REVIEW

**Comment by NiklasGustafsson** [comment]
@tarekgh -- this is massive, but most of the changes follow exactly the same pattern. The most important changes to review are in Module.cs and in ScriptModule.cs, the rest are pretty templated.

@lostmsu -- if you could give it a quick check, that would be great, but don't feel obligated to. It's the "ease of use" aspect of the fix to the issue you reported, not the rewrite of the modules in C#.

File: src/TorchSharp/JIT/ScriptModule.cs [hunk 1]
```diff
@@ -163,9 +163,63 @@ public override void eval()
...
+                            var value = field.GetValue(this);
+
+                            switch (value) {
+                            // This test must come before the Tensor test
+                            case Modules.Parameter param when dtype == param.dtype && device.type == param.device_type && device.index == param.device_index:
+                                continue;
+
+                            case Modules.Parameter param: {
+                                    var t = param.to(dtype, device);
+                                    t.retain_grad();
```
Comment (lostmsu) [inline src/TorchSharp/JIT/ScriptModule.cs:198]: Why is this needed? Should it only be called when the original parameter `requires_grad`?

File: src/TorchSharp/JIT/ScriptModule.cs [hunk 2]
```diff
@@ -163,9 +163,63 @@ public override void eval()
...
+                                    var t = tensor.to(dtype, device);
+                                    field.SetValue(this, t);
+                                    ConditionallyRegisterBuffer(fieldName, t);
+                                    break;
+                                }
+                            }
+                        }
+
+                        _deviceType = device.type;
+                        _deviceIndex = device.index;
```
Comment (lostmsu) [inline src/TorchSharp/JIT/ScriptModule.cs:215]: Not related to this change, but is having device info cached in the object safe? Can I obtain two C# objects pointing to the same module? If so, and I move the module using one object, the other's state becomes inconsistent?

File: src/TorchSharp/NN/Module.cs [hunk 3]
```diff
@@ -142,12 +143,66 @@ protected virtual void Dispose(bool disposing)
...
+                [DllImport("LibTorchSharp")]
+                static extern void THSNN_Module_to_device_dtype(HType module, sbyte dtype, long deviceType, long deviceIndex);
+
                 [DllImport("LibTorchSharp")]
                 static extern void THSNN_Module_to_device(HType module, long deviceType, long deviceIndex);
 
                 [DllImport("LibTorchSharp")]
                 static extern void THSNN_Module_to_dtype(HType module, sbyte dtype);
 
+                public virtual Module to(Device device, ScalarType dtype)
```
Comment (lostmsu) [inline src/TorchSharp/NN/Module.cs:155]: Haven't I seen this exact method body before. Can it be reused?

File: src/TorchSharp/NN/Module.cs [hunk 4]
```diff
@@ -980,6 +1035,30 @@ protected virtual void RegisterComponents()
...
+                    if (!dtype.HasValue)
+                        dtype = get_default_dtype();
+
+                    if (device == null)
+                        device = torch.CPU;
+
+                    return (device, dtype.Value);
+                }
+
+                internal T MoveModule<T>(Device device, ScalarType? dtype) where T : Module
```
Comment (lostmsu) [inline src/TorchSharp/NN/Module.cs:1049]: Make this a static extension method, then C# will infer the return type.

File: test/TorchSharpTest/TestTorchTensorBugs.cs [hunk 5]
```diff
@@ -531,8 +531,8 @@ public Module510(int in_channels, int out_channels, int kernel_size = 3, int str
                     ReLU(inplace: true)
                 );
 
-                temp.weight = Parameter(torch.randn(temp.weight.shape));
-                temp.bias = Parameter(torch.randn(temp.bias.shape));
+                temp.weight = Parameter(torch.randn(temp.weight!.shape));
```
Comment (lostmsu) [inline test/TorchSharpTest/TestTorchTensorBugs.cs:534]: Can you actually null `weight` of `BatchNorm1d`? Is it a legal operation that does not break things?

Comment (NiklasGustafsson) [reply test/TorchSharpTest/TestTorchTensorBugs.cs:534]: PyTorch allows it. Not sure what happens after...

File: src/TorchSharp/NN/Module.cs (hunk 3 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs:160]: The ScriptModule version is an override that calls to a different native method than non-script modules do.

File: src/TorchSharp/JIT/ScriptModule.cs (hunk 2 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/JIT/ScriptModule.cs:215]: That needs to be investigated. As you say, not related to this change, specifically.

File: src/TorchSharp/NN/Module.cs (hunk 4 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs:1052]: I'll take a look at that.

File: src/TorchSharp/JIT/ScriptModule.cs (hunk 1 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/JIT/ScriptModule.cs:198]: Parameters should always require gradients.

File: src/TorchSharp/NN/Module.cs (hunk 3 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs:160]: ... but there are still opportunities for sharing. Thanks for pointing that out.

Comment (tarekgh) [reply src/TorchSharp/NN/Module.cs:160]: Right, I think there is no need to override this method in SxriptModule.

File: src/TorchSharp/NN/Module.cs [hunk 6]
```diff
@@ -142,12 +143,74 @@ protected virtual void Dispose(bool disposing)
...
+                                field.SetValue(this, p);
+                                ConditionallyRegisterParameter(fieldName, p);
+                                break;
+                            }
+
+                        case Tensor tensor when (device.type != tensor.device_type || device.index != tensor.device_index): {
+                                var t = tensor.to(dtype, device);
+                                field.SetValue(this, t);
+                                ConditionallyRegisterBuffer(fieldName, t);
+                                break;
```
Comment (tarekgh) [inline src/TorchSharp/NN/Module.cs:202]: nit: please align this block.

File: src/TorchSharp/NN/Module.cs [hunk 7]
```diff
@@ -142,12 +143,74 @@ protected virtual void Dispose(bool disposing)
...
+                    InitializeDeviceType(device.type);
+                    THSNN_Module_to_device_dtype(handle, (sbyte)dtype, (int)device.type, device.index);
+                    CheckForErrors();
+
+                    _toEpilog(device, dtype);
+
+                    return this;
+                }
+
+                protected void _toEpilog(Device device, ScalarType dtype)
```
Comment (tarekgh) [inline src/TorchSharp/NN/Module.cs:175]: >_toEpilog [](http://example.com/codeflow?start=31&length=9)

The code here is identical to the code in the same method that takes DeviceType. Is there a way we can share the code instead of duplicating it? Or we can make this method a generic method?

File: src/TorchSharp/NN/Module.cs [hunk 8]
```diff
@@ -980,6 +1053,30 @@ protected virtual void RegisterComponents()
...
+                internal T MoveModule<T>(Device device, ScalarType? dtype) where T : Module
+                {
+                    T module = (T)this;
+                    if (device != null && dtype.HasValue) {
+                        module = (T)module.to(device, dtype.Value);
+                    } else if (device != null) {
+                        module = (T)module.to(device);
+                    } else if (dtype.HasValue) {
+                        module = (T)module.to(dtype.Value);
+                    }
```
Comment (tarekgh) [inline src/TorchSharp/NN/Module.cs:1076]: nit:

```C#
return device != null ? 
           (dtype.HasValue ? (T)module.to(device, dtype.Value) : (T)module.to(device)) :
           (dtype.HasValue ? (T)module.to(dtype.Value) : module);
````

**Review by tarekgh — APPROVED** [review]
